indra/newview/llvoicewebrtc.cpp

```diff
@@ -2070,123 +2104,92 @@ void LLVoiceWebRTCConnection::OnIceCandidate(const llwebrtc::LLWebRTCIceCandidat
-                    for (auto &ice_candidate : mIceCandidates)
-                    {
-                        LLSD body_candidate;
-                        body_candidate["sdpMid"]        = ice_candidate.mSdpMid;
-                        body_candidate["sdpMLineIndex"] = ice_candidate.mMLineIndex;
-                        body_candidate["candidate"]     = ice_candidate.mCandidate;
-                        candidates.append(body_candidate);
-                    }
-                    body["candidates"] = candidates;
-                    mIceCandidates.clear();
-                }
-                else if (mIceCompleted)
-                {
-                    LLSD body_candidate;
-                    body_candidate["completed"] = true;
-                    body["candidate"]           = body_candidate;
-                    iceCompleted                = mIceCompleted;
-                    mIceCompleted               = false;
-                }
+        LLCoreHttpUtil::HttpCoroutineAdapter::ptr_t httpAdapter(
+            new LLCoreHttpUtil::HttpCoroutineAdapter("LLVoiceWebRTCAdHocConnection::processIceUpdatesCoro",
+                                                        LLCore::HttpRequest::DEFAULT_POLICY_ID));
+        LLCore::HttpRequest::ptr_t httpRequest(new LLCore::HttpRequest);
+        LLCore::HttpOptions::ptr_t httpOpts(new LLCore::HttpOptions);
 
-                body["viewer_session"]    = mViewerSession;
-                body["voice_server_type"] = WEBRTC_VOICE_SERVER_TYPE;
-
-                LLCore::HttpRequest::policy_t               httpPolicy(LLCore::HttpRequest::DEFAULT_POLICY_ID);
-                LLCoreHttpUtil::HttpCoroutineAdapter::ptr_t httpAdapter(
-                    new LLCoreHttpUtil::HttpCoroutineAdapter("voiceAccountProvision", httpPolicy));
-                LLCore::HttpRequest::ptr_t httpRequest(new LLCore::HttpRequest);
-                LLCore::HttpOptions::ptr_t httpOpts = LLCore::HttpOptions::ptr_t(new LLCore::HttpOptions);
-                LLCoreHttpUtil::HttpCoroutineAdapter::callbackHttpPost(
-                    url,
-                    LLCore::HttpRequest::DEFAULT_POLICY_ID,
-                    body,
-                    boost::bind(&LLVoiceWebRTCSpatialConnection::onIceUpdateComplete, this, iceCompleted, _1),
-                    boost::bind(&LLVoiceWebRTCSpatialConnection::onIceUpdateError, this, 3, url, body, iceCompleted, _1));
-                mOutstandingRequests++;
-                mTrickling = true;
-            }
+        httpOpts->setWantHeaders(true);
+
+        LLSD result = httpAdapter->postAndSuspend(httpRequest, url, body, httpOpts);
+        LLSD httpResults = result[LLCoreHttpUtil::HttpCoroutineAdapter::HTTP_RESULTS];
+        LLCore::HttpStatus status = LLCoreHttpUtil::HttpCoroutineAdapter::getStatusFromLLSD(httpResults);
+
+        if (!status)
+        {
+            // couldn't trickle the candidates, so restart the session.
+            setVoiceConnectionState(VOICE_STATE_SESSION_RETRY);
```

Comment:
postAndSuspend can take undetermined amount of time and coroutine can persit till end of shutdown. At this point LLVoiceWebRTCConnection might no longer exists.

In case of coroutines you need to check if class is still alive in some way after any 'suspend' (ex: call checkStop(), not sure why suspend doesn't call it internally after a yield)

Reply:
Same issue with breakVoiceConnectionCoro

Reply:
thanks, will fix

Reply:
LLVivoxVoiceClient::sShuttingDown exists for the same reason - vivox's corouitine stubbornly refuses to die.

Reply:
Used sShuttingDown in LLWebRTCVoiceClient.  Should be good now.